modules/ldap/src/main/java/com/neighborhoods/wasabi/ldap/impl/LdapUserDirectory.java

```diff
@@ -0,0 +1,240 @@
+     * @return true, if a directory token is valid
+     */
+    @Override
+    public boolean isDirectoryTokenValid(String username, String encryptedPassword) {
+        return delegate.isDirectoryTokenValid(this, username, encryptedPassword);
+    }
+
+    /**
+     * Find a user by username
+     * 
+     * @param username The target user's username
+     * @return The user object
+     */
+    @Override
+    public DirectoryUser lookupDirectoryUser(String username) {
+        DirectoryUser result = getDirectoryUserFromCache(username);
+        if (result != null) {
+            return result;
+        }
+        return delegate.getDirectoryUserByUsername(this, username);
+    }
+
+    /**
+     * Find a user by username
+     * 
+     * @param username The username of the target user
+     * @return The resulting user object
+     */
+    @Override
+    public UserInfo lookupUser(final Username username) {
+        DirectoryUser result = getDirectoryUserFromCache(username.getUsername());
+        if (result != null) {
+            return result.getUserInfo();
+        }
+        result = delegate.getDirectoryUserByUsername(this, username.getUsername());
+        if (result != null) {
+            return result.getUserInfo();
+        }
+        return null;
+    }
+
+    /**
+     * Lookup a user by email
+     * 
+     * @param userEmail the target user's email
+     * @return The resulting user object
+     */
+    @Override
+    public UserInfo lookupUserByEmail(final String userEmail) {
+        DirectoryUser result = null;
+        Optional<DirectoryUser> matchingUser = getAllUsers().filter(user -> user.getEmail().equals(userEmail))
+                .findFirst();
+        if (matchingUser.isPresent()) {
+            result = validateCachedUser(matchingUser.get());
+        }
+        if (result == null) {
+            result = delegate.getDirectoryUserByEmail(this, userEmail);
+        }
+        if (result != null) {
+            return result.getUserInfo();
+        }
+        throw new AuthenticationException(format("Email address does not exist: {0}", userEmail));
+    }
+
+    /**
+     * Refresh the cache
+     */
+    @Override
+    public void refreshCache() {
+        if (this.users == null) {
+            this.users = new ConcurrentHashMap<String, DirectoryUser>();
+        } else {
+            this.users.clear();
+        }
+        delegate.populateUserCache(this);
+    }
+
+    /**
+     * Remove a user from the cache
+     * 
+     * @param The username of the user to remove
+     * @return The result of the remove action
+     */
+    @Override
+    public DirectoryUser removeUserFromCache(String username) {
+        return this.users.remove(username);
+    }
+
+    /**
+     * Validate a cached user (check expiration)
+     *
+     * @param cachedUser the cached user
+     * @return the directory user
+     */
+    protected DirectoryUser validateCachedUser(DirectoryUser cachedUser) {
```

Comment:
i feel like it's a bit weird for this to return a `DirectoryUser`
I'd either expect a boolean, or if you follow the pattern the wasabi interfaces have, `void` and throwing an exception if invalid.

that said, it is not a strong objection

Reply:
yea i'm on board with retaining the signature/contract and relying on naming for clarity.
@nafarius333's option is pretty good.
the current name actually works ok if you understand "validation" as a verb acting on and potentially mutating the `cachedUser`, so i'm ok with it too.